Drop Pulpcore 3.14 & 3.15, move to 3.16 and 3.17

ekohl opened the pull request.

ekohl commented on spec/acceptance/basic_spec.rb:
```diff
@@ -65,8 +65,7 @@
   describe curl_command("https://#{host_inventory['fqdn']}/pulp/api/v3/", cacert: "#{certdir}/ca-cert.pem") do
-    its(:response_code) { is_expected.to eq(200) }
-    its(:body) { is_expected.not_to contain('artifacts_list') }
+    its(:response_code) { is_expected.to eq(403) }
```
I do wonder if this is correct. Also, why is the HTTP 200 code in the other tests (plugin, content route spec).

evgeni replied:
Doesn't look correct to me :)

evgeni replied:
On the other hand, what it this supposed to be? The api root, listing all the endpoints? Does it work if you pass a cert/key?

evgeni replied:
I *think* it's a side-effect of https://github.com/pulp/pulpcore/pull/1593 (and I've asked in pulp-dev to make sure).

If you want to test an endpoint w/o auth, use `/pulp/api/v3/status/` as we do in line 57.

That said, I am not really sure what this test is supposed to test. It was added as part of https://github.com/theforeman/puppet-pulpcore/commit/a60df532a95dbb3bf951aaf10b85f9b31ba3e635 where we implemented HTTPS handling inside this module, but it isn't saying what it's testing :(

ekohl replied:
The idea was mostly that it showed a difference between authenticated and non-authenticated. With that you can see that certificate authentication works. That's something this module sets up. If you have a better idea of how to test auth, I'm open to suggestions.

evgeni replied:
But that's exactly what it did not test, as access to `/` was explicitly allowed without auth (just the content differed).

I would suggest hitting a *real* API endpoint like `/users` or `/artifacts` or `/tasks`. Without auth you get status=403 and a JSON reply of `{"detail":"Authentication credentials were  provided."}`, with you get status=200 and some actual content.

ekohl replied:
> But that's exactly what it did not test, as access to `/` was explicitly allowed without auth (just the content differed).

I disagree it did not test it. By checking the difference in output I believe I was able to test it. You may be right that it happened to be a side effect. I'd be happy to merge a follow up PR.

evgeni replied:
https://github.com/theforeman/puppet-pulpcore/pull/251

ekohl commented:
The failures look like bugs in Pulp 3.17:
```
Info: Pulpcore::Admin[migrate --noinput]: Scheduling refresh of Exec[pulpcore-manager migrate --noinput]
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: /opt/theforeman/tfm-pulpcore/root/usr/lib64/python3.8/site-packages/cryptography/hazmat/bindings/openssl/binding.py:173: CryptographyDeprecationWarning: OpenSSL version 1.0.2 is no longer supported by the OpenSSL project, please upgrade. The next version of cryptography will drop support for it.
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   warnings.warn(
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: Traceback (most recent call last):
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/theforeman/tfm-pulpcore/root/usr/lib/python3.8/site-packages/git/__init__.py", line 87, in <module>
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     refresh()
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/theforeman/tfm-pulpcore/root/usr/lib/python3.8/site-packages/git/__init__.py", line 76, in refresh
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     if not Git.refresh(path=path):
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/theforeman/tfm-pulpcore/root/usr/lib/python3.8/site-packages/git/cmd.py", line 340, in refresh
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     raise ImportError(err)
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: ImportError: Bad git executable.
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: The git executable must be specified in one of the following ways:
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     - be included in your $PATH
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     - be set via $GIT_PYTHON_GIT_EXECUTABLE
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     - explicitly set via git.refresh()
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: 
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: All git commands will error until this is rectified.
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: 
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: This initial warning can be silenced or aggravated in the future by setting the
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: $GIT_PYTHON_REFRESH environment variable. Use one of the following values:
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     - quiet|q|silence|s|none|n|0: for no warning or exception
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     - warn|w|warning|1: for a printed warning
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     - error|e|raise|r|2: for a raised exception
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: 
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: Example:
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     export GIT_PYTHON_REFRESH=quiet
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: 
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: 
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: The above exception was the direct cause of the following exception:
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: 
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: Traceback (most recent call last):
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/theforeman/tfm-pulpcore/root/usr/bin/pulpcore-manager", line 11, in <module>
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     load_entry_point('pulpcore==3.17.3', 'console_scripts', 'pulpcore-manager')()
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/theforeman/tfm-pulpcore/root/usr/lib/python3.8/site-packages/pulpcore/app/manage.py", line 11, in manage
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     execute_from_command_line(sys.argv)
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/theforeman/tfm-pulpcore/root/usr/lib/python3.8/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     utility.execute()
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/theforeman/tfm-pulpcore/root/usr/lib/python3.8/site-packages/django/core/management/__init__.py", line 395, in execute
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     django.setup()
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/theforeman/tfm-pulpcore/root/usr/lib/python3.8/site-packages/django/__init__.py", line 24, in setup
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     apps.populate(settings.INSTALLED_APPS)
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/theforeman/tfm-pulpcore/root/usr/lib/python3.8/site-packages/django/apps/registry.py", line 122, in populate
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     app_config.ready()
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/theforeman/tfm-pulpcore/root/usr/lib/python3.8/site-packages/pulpcore/app/apps.py", line 96, in ready
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     self.import_viewsets()
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/theforeman/tfm-pulpcore/root/usr/lib/python3.8/site-packages/pulpcore/app/apps.py", line 140, in import_viewsets
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     self.viewsets_module = import_module(viewsets_module_name)
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/rh/rh-python38/root/usr/lib64/python3.8/importlib/__init__.py", line 127, in import_module
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     return _bootstrap._gcd_import(name[level:], package, level)
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/theforeman/tfm-pulpcore/root/usr/lib/python3.8/site-packages/pulp_ansible/app/viewsets.py", line 33, in <module>
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     from pulp_ansible.app.galaxy.mixins import UploadGalaxyCollectionMixin
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/theforeman/tfm-pulpcore/root/usr/lib/python3.8/site-packages/pulp_ansible/app/galaxy/mixins.py", line 3, in <module>
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     from pulp_ansible.app.tasks.collections import import_collection
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/theforeman/tfm-pulpcore/root/usr/lib/python3.8/site-packages/pulp_ansible/app/tasks/collections.py", line 72, in <module>
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     from git import GitCommandError, Repo
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:   File "/opt/theforeman/tfm-pulpcore/root/usr/lib/python3.8/site-packages/git/__init__.py", line 89, in <module>
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     raise ImportError('Failed to initialize: {0}'.format(exc)) from exc
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: ImportError: Failed to initialize: Bad git executable.
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: The git executable must be specified in one of the following ways:
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     - be included in your $PATH
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     - be set via $GIT_PYTHON_GIT_EXECUTABLE
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     - explicitly set via git.refresh()
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: 
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: All git commands will error until this is rectified.
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: 
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: This initial warning can be silenced or aggravated in the future by setting the
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: $GIT_PYTHON_REFRESH environment variable. Use one of the following values:
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     - quiet|q|silence|s|none|n|0: for no warning or exception
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     - warn|w|warning|1: for a printed warning
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     - error|e|raise|r|2: for a raised exception
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: 
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns: Example:
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[pulpcore-manager migrate --noinput]/returns:     export GIT_PYTHON_REFRESH=quiet
  Error: 'pulpcore-manager migrate --noinput' returned 1 instead of one of [0]
```

ekohl commented:
Updated the PR to only 3.16 since 3.17 is broken on EL7.

ekohl commented:
Note to self: include enabling the pulpcore el8 module in `repo.pp`.

ehelms commented on manifests/repo.pp:
```diff
@@ -19,6 +20,15 @@
 
+  package { 'pulpcore-dnf-module':
+    name        => 'pulpcore',
+    ensure      => $dist_tag,
```
```suggestion
    ensure      => $dist_tag,
    name        => 'pulpcore',
```

evgeni commented on spec/classes/repo_spec.rb:
```diff
@@ -10,7 +10,14 @@
           .with_content(%r{^baseurl=https://yum.theforeman.org/pulpcore/\d+\.\d+/el\d+/\$basearch$})
-          .that_comes_before('Anchor[pulpcore::repo]')
+          .that_notifies('Anchor[pulpcore::repo]')
+        is_expected.to contain_package('pulpcore-dnf-module')
```
this is currently expected on EL7 too, leading to failures:

```
pulpcore::repo on centos-7-x86_64 is expected to contain Package[pulpcore-dnf-module] that requires File[/etc/yum.repos.d/pulpcore.repo]
```

ekohl replied:
Yes, trying to quickly do this before the next meeting is proving to be not the best strategy.

evgeni commented on manifests/repo.pp:
```diff
@@ -19,6 +20,17 @@
     notify  => Anchor['pulpcore::repo'],
   }
 
+  if $facts['os']['release']['major'] == '8' {
```
```suggestion
  if $facts['os']['release']['major'] != '7' {
```

to match tests? (and the idea of future proofing) :)

ekohl replied:
I copied it from here: https://github.com/theforeman/puppet-foreman/blob/9e483426c511e7a7a2648c2eaba18f50ac7df8f9/manifests/repo.pp#L44
But you're right.

ekohl commented:
Updated:
* Added 3.17 again (and made it the default)
* Addressed https://github.com/theforeman/puppet-pulpcore/pull/249#discussion_r829057486
* Included the test from https://github.com/theforeman/puppet-pulpcore/pull/251

evgeni commented on README.md:
```diff
@@ -10,15 +10,13 @@ All supported versions are listed below. For every supported version, acceptance
 Supported operating systems are listed in `metadata.json` but individual releases can divert from that. For example, if Pulpcore x.y drops EL7, it will still be listed in metadata.json until all versions supported by the module have dropped it. Similarly, if x.z adds support for EL9, it'll be listed in `metadata.json` and all versions that don't support EL9 will have a note.
 
-### Pulpcore 3.15
+### Pulpcore 3.16
```
```suggestion
### Pulpcore 3.17
```

;-)

Even tho I am not sure whether 3.16 or 3.17 should be the default. At least today Katello doesn't use 3.17 yet.

ekohl replied:
I'd say default to the newest is good, even if Katello isn't using it.

evgeni replied:
wfm
